### PR TITLE
initial support for corpora schema conventions

### LINKED ISSUE #1532: corpora dataset project info populated into CXG for display in project/dataset card

### PR DESCRIPTION
**Pull request by bkmartinjr**
This PR adds:
* specification for CXG, including CXG encoding of Corpora schema conventions
* add support to extract Corpora dataset props, and serve them to the front-end via /config route
* cxgtool.py - add support to extract and encode Corpora dataset props in CXG

This is a partial solution to #1509 
Fixes #1532

### CODE REVIEW

File: server/common/corpora.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        corpora_schema_version is not None
+        and corpora_encoding_version is not None
+        and corpora_schema_version.startswith("1.")
+        and corpora_encoding_version.startswith("0.1.")
+    )
+
+    if not version_is_supported:
+        raise ValueError("Unsupported Corpora schema version")
+
+    Required_Simple_Fields = [
```
Comment (maniarathi) [inline server/common/corpora.py:55]: Just a nit: why is the casing weird on these variables?

File: server/common/corpora.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    Get Corpora dataset properties from an AnnData
+    """
+    versions = corpora_get_versions_from_anndata(adata)
+    if versions is None:
+        return None
+    [corpora_schema_version, corpora_encoding_version] = versions
+    version_is_supported = (
+        corpora_schema_version is not None
+        and corpora_encoding_version is not None
+        and corpora_schema_version.startswith("1.")
```
Comment (maniarathi) [inline server/common/corpora.py:48]: Can we pull out the supported schema versions into a utility file or some sort of constants file. I see that `cxgtool.py` also checks the versions so it would be helpful to have the supported versions in a single location that can get updated instead of having to remember a couple places to check.

File: server/test/test_diffexp.py [hunk 3]
```diff
@@ -105,13 +105,23 @@ def sparse_diffexp(self, apply_col_shift):
             adata = adaptor_anndata.data
 
             sparsename = os.path.join(dirname, "sparse.cxg")
-            write_cxg(adata=adata, container=sparsename, title="sparse", sparse_threshold=11)
+            cxg_group_metadata = create_cxg_group_metadata(
```
Comment (maniarathi) [inline server/test/test_diffexp.py:108]: Do we also need a test case here where `corpora_props` is not None since that is new functionality?

File: dev_docs/cxg.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+## Entities
+
+### CXG
+
+The CXG is a TileDB group containing all data and metadata for a single annotated matrix.  The following objects must be present in a CXG:
+* __obs__: a TileDB array, of shape (n_obs,), containing obs annotations, each annotation stored in a separate TileDB array attribute.
+* __var__: a TileDB array, of shape (n_var,), containing var annotations, each annotation stored in a separate TileDB array attribute.
+* __X__: a TileDB array, of shape (n_obs, n_var), with a single TileDB attribute of numeric type.
+* __X_col_shift__: (optional) TilebDB Array used in column shift encoding, shape (n_var,), dtype = X.dtype. Single unnamed numeric attribute.
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply dev_docs/cxg.md:47]: @bmccandless - please add this information to this branch.

File: server/common/corpora.py (hunk 1 above)
Comment (bkmartinjr) [reply server/common/corpora.py:55]: casing?

File: server/common/corpora.py (hunk 2 above)
Comment (bkmartinjr) [reply server/common/corpora.py:48]: Can you expand on what problem this solves?  I bias away from overly complicated config & code structure unless there is a clear value.   Right now, this file is entirely self-contained wrt semantics and readability.  Given how simple the code is, I'm not sure splitting helps.

LMK.

File: dev_docs/cxg.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        "version": {
+            "corpora_schema_version": "1.0.0",
+            "corpora_encoding_version": "0.1.0",
+        }
+    }
+}
+```
+
+The `corpora` metadata, if present, MUST contain the version information. Optionality of other values in this object will follow the specifications set forth in the relevant Corpora schema specification (ie, optional fields are optional, required are present, etc), with the following changes:
+* the contents of `corpora_encoding_version` MUST be identical to the `cxg_version`, as this field is defined as the current object encoding version, *NOT* the source data encoding version.
```
Comment (bmccandless) [inline dev_docs/cxg.md:138]: If corpora_encoding_version must be identical to cxg_version, then why have it?

File: server/test/test_diffexp.py (hunk 3 above)
Comment (bkmartinjr) [reply server/test/test_diffexp.py:108]: @bmccandless - can you respond.  I'm unclear why this parameter would be necessary to test for differential expression.

File: dev_docs/cxg.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+All other Corpora schema fields will be encoded into a CXG using the conventions defined in the [Corpora Schema AnnData Implementation](https://github.com/chanzuckerberg/corpora-data-portal/blob/main/backend/schema/corpora_schema_h5ad_implementation.md). For example, fields in `AnnData.obs` will be encoded in the CXG `obs`array as defined [above](#obs-and-var).
+
+### Compatibility with CXG 0.1.0
+
+For backwards compatibility and continuity with CXG version 0.1.0, the following MUST be implemented.
+
+#### Presentation Hints
+* The [Corpora `title`](https://github.com/chanzuckerberg/corpora-data-portal/blob/main/backend/schema/corpora_schema.md#presentation-metadata) value MUST be saved in the `cxg_properties.title` field.
+* The [Corpora `color_map`](https://github.com/chanzuckerberg/corpora-data-portal/blob/main/backend/schema/corpora_schema.md#presentation-hints) MUST be saved in the `cxg_category_colors` field.
```
Comment (bmccandless) [inline dev_docs/cxg.md:152]: I think is is listed as optional in the corpora schema.  "Submitters can include a field called {field}_colors..."

File: server/common/corpora.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+from server.cli.upgrade import validate_version_str
+
+
+def corpora_get_versions_from_anndata(adata):
+    """
+    Given an AnnData object, return:
+        * None - if not a Corpora object
+        * [ corpora_schema_version, corpora_encoding_version ] - if a Corpora object
+
+    Implements the identification protocol defined in the spect.
```
Comment (bmccandless) [inline server/common/corpora.py:19]: s/spect/spec/

File: server/common/corpora.py (hunk 2 above)
Comment (maniarathi) [reply server/common/corpora.py:48]: I figured that because this particular check is written word-for-word twice (lines 551-556) in a completely different file `cxgtool.py`, it seems preferable to consolidate into a single location. Otherwise you run into the issue of having to remember to update in both locations or run into bugs.

File: server/common/corpora.py (hunk 1 above)
Comment (maniarathi) [reply server/common/corpora.py:55]: The capitalization of the beginning of each word threw me off :P Haven't seen it before and didn't know what it meant.

Comment (bkmartinjr) [reply server/common/corpora.py:55]: lolz, no idea - just me trying to imply constants.  I'll remove it on the assumption you are the person who defines code convensions.

File: server/common/corpora.py (hunk 2 above)
Comment (bkmartinjr) [reply server/common/corpora.py:48]: Ah, I see - you think both files will always have the same version check.  I was not assuming that.  But for now, I'll DRY it.

File: dev_docs/cxg.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bmccandless) [reply dev_docs/cxg.md:47]: There is wording in the next section that explains each of these arrays in more detail.  If looks like this section is just meant to be a brief summary of the arrays.

[the remark this answers is not in the document]
Comment (bkmartinjr) [reply dev_docs/cxg.md:47]: correct - but I assume @maniarathi wants even more details than the following section?  Arathi?

File: dev_docs/cxg.md (hunk 5 above)
Comment (bkmartinjr) [reply dev_docs/cxg.md:138]: Because cxg_version is a super set.   They evolve independently.

File: dev_docs/cxg.md (hunk 6 above)
Comment (bkmartinjr) [reply dev_docs/cxg.md:152]: correct, but when it is present, it must be int that field, rather than in the corpora field.

I'll revise the wording to make it more obvious what I'm trying to say.

File: server/test/test_diffexp.py (hunk 3 above)
Comment (bmccandless) [reply server/test/test_diffexp.py:108]: This test is just to verify that the sparse matrix converter works.   
A test case for corpora_props should probably go in the test_cxgtool.py

File: dev_docs/cxg.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bmccandless) [reply dev_docs/cxg.md:47]: @maniarathi, there is no link to col shift encoding.  It is our own invention to recapture the inherit sparseness of a matrix that had been previously normalized.   The cxgtool.py automatically can detect this and do the transformation.   A user would not need to know about it (but the details are pretty simple).

File: server/test/test_diffexp.py (hunk 3 above)
Comment (bkmartinjr) [reply server/test/test_diffexp.py:108]: write_cxg is not a public interface, and will never have a None value in props.  Matt and Brian pulled  this out simply for testing.

I will add a test case for the corpora.py exported API.

Comment (bkmartinjr) [reply server/test/test_diffexp.py]: added test_corpora.py

File: server/common/corpora.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    ]
+    # Spec says some values encoded as JSON due to the inability of AnnData to store complex types.
+    required_json_fields = ["contributors", "project_links"]
+    optional_simple_fields = ["preprint_doi", "publication_doi", "default_embedding", "default_field", "tags"]
+
+    corpora_props = {}
+    for key in required_simple_fields:
+        if key not in adata.uns:
+            raise KeyError(f"missing Corpora schema field {key}")
+        corpora_props.update({key: adata.uns[key]})
```
Comment (mckinsel) [inline server/common/corpora.py:78]: This is a kind of unusual way to assign to a dict. Any reason not to `corpora_props[key] = adata.uns[key]`?

File: server/common/corpora.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    for key in required_json_fields:
+        if key not in adata.uns:
+            raise KeyError(f"missing Corpora schema field {key}")
+        corpora_props.update({key: json.loads(adata.uns[key])})
```
Comment (mckinsel) [inline server/common/corpora.py:82]: You might want to add a little color to a `json.JSONDecodeError` here, like `try: ... except ...: raise json.JSONDecodeError(f"Corpora schema field {key} is expected to be a valid JSON string")`

File: server/converters/cxgtool.py [hunk 10]
```diff
@@ -601,6 +545,59 @@ def save_metadata(container, metadata_dict):
...
+        # clobber encoding version to be OUR version, not the source H5AD encoding
+        corpora_props["version"].update({"corpora_encoding_version": CXG_VERSION})
+        corpora_project_links = corpora_props.get("project_links", [])
+        corpora_about_link = next(
+            (link for link in corpora_project_links if (link.get("link_type", None) == "SUMMARY")), {}
+        )
+    else:
+        corpora_about_link = {}
+
+    title = title if title is not None else corpora_about_link.get("link_name", basefname)
```
Comment (mckinsel) [inline server/converters/cxgtool.py:579]: This is kind of a personal preference, so take it or leave it, but

```python
title = title or corpora_about_link.get("link_name", basefname)
```

File: dev_docs/cxg.md [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+All implementions MUST ignore unrecognized fields. 
+
+#### cxg_category_colors
+
+This optional field contains a copy of the category color table, which MAY be used to display category-specific color labels. This is a JSON dictionary, containing a per-category color-table.  Each color table is named `{category_name}_colors`, and is itself a dictionary mapping label name to RGB color.  For example:
+
+```
+{
+    "louvain_colors": {
+        "0": "#FFFFFF",
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply dev_docs/cxg.md:112]: It currently supports several formats, but that is really an app convention (currently unspecified), and I'm not sure if it belongs in this spec or not.  At the moment, it will support either of:
* rgb(DDD,DDD,DDD) -- eg, `rgb(255,255,255)` (base 10)
* #HHHHHH -- eg, `#FF0000` (hex)

File: server/common/corpora.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    """
+    Given an AnnData object, return:
+        * None - if not a Corpora object
+        * [ corpora_schema_version, corpora_encoding_version ] - if a Corpora object
+
+    Implements the identification protocol defined in the specification.
+    """
+
+    # per Corpora AnnData spec, this is a corpora file if the following is true
+    if adata.uns is None or "version" not in adata.uns:
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/common/corpora.py:24]: per the [AnnData spec](https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData), `None` is legal, and is the default value.

File: server/common/corpora.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        return None
+    version = adata.uns["version"]
+    if not isinstance(version, collections.abc.Mapping) or "corpora_schema_version" not in version:
+        return None
+
+    corpora_schema_version = version.get("corpora_schema_version", None)
+    corpora_encoding_version = version.get("corpora_encoding_version", None)
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/common/corpora.py:31]: true, I was just being explicitly explicit :-).   Will remove.

File: server/common/corpora.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+    # TODO: spec says these must be SEMVER values, so check.
+    if validate_version_str(corpora_schema_version) and validate_version_str(corpora_encoding_version):
+        return [corpora_schema_version, corpora_encoding_version]
+
+    return None
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/common/corpora.py:37]: Can you clarify?   Do you mean returning None has no side-effects?  Or?

File: server/common/corpora.py (hunk 8 above)
Comment (bkmartinjr) [reply server/common/corpora.py:78]: I simply lack any style sensibility :-)  Will fix.

File: dev_docs/cxg.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (mckinsel) [reply dev_docs/cxg.md:112]: Okay thanks, was mostly just curious.

File: server/common/corpora.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mckinsel) [reply server/common/corpora.py:24]: I think `None` is the default parameter value in the constructor, but I'm pretty sure it [always sticks something in there](https://github.com/theislab/anndata/blob/master/anndata/_core/anndata.py#L506).

File: server/common/corpora.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (mckinsel) [reply server/common/corpora.py:37]: If you delete that line the behavior of the function does not change:
```python
>>> def fn():
        if False: return "something"
>>> fn() is None
True
```
Not a big deal, if you think it's clearer with the explicit `return`.

File: server/common/corpora.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/common/corpora.py:24]: I have previously been bit by not following their documentation (as they feel free to change the code on occasion).  Unless you see a major advantage, I'd prefer to just assume their docs are correct, even if it costs us an extra check....

File: server/common/corpora.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/common/corpora.py:37]: doh...got it.

File: server/common/corpora.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mckinsel) [reply server/common/corpora.py:24]: Extremely possible I'm missing something, but I don't think the docs say that `adata.uns` may be `None`? The docs don't appear to express a super crisp opinion about the question though. Maybe the defensive thing to do is `if "version" not in adata.uns_keys()`?

[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/common/corpora.py:24]: The `uns` field definition on that page is:  `uns : Optional[Mapping[str, Any]] (default: None)`

This is confusing, as `None` is not a Mapping.  Hence my conservatism.

So yeah, your approach seems good.
